src/main/java/io/dockstore/provision/DOSPlugin.java

```diff
@@ -49,24 +56,63 @@ public void stop() {
     @Extension
     public static class DOSPreProvision implements PreProvisionInterface {
 
-        DOSPluginUtil dosPluginUtil = new DOSPluginUtil();
         static final Set<String> SCHEME = new HashSet<>(Lists.newArrayList("dos"));
+        static final String SCHEME_PREFERENCE = "scheme-preference";
+        static final int GET_SCHEME = 0;
+        static final String PROTOCOL = ":\\/\\/(.+)/";
+
+        DOSPluginUtil dosPluginUtil = new DOSPluginUtil();
+        List<String> preferredSchemes = new ArrayList<>();
+
+        @Override
+        public void setConfiguration(Map<String, String> map) {
+            MapConfiguration config = new MapConfiguration(map);
+            config.setListDelimiterHandler(new DefaultListDelimiterHandler(','));
+
+            // Retrieve the scheme preferences from the config file
+            this.preferredSchemes = config.getList(String.class, SCHEME_PREFERENCE, Collections.emptyList());
+            // Remove any empty strings from the list
+            this.preferredSchemes.removeIf(e -> e.equals(""));
+        }
 
         public Set<String> schemesHandled() {
             return SCHEME;
         }
 
         public List<String> prepareDownload(String targetPath) {
             List<String> urlList = new ArrayList<>();
+
+            // Linked Hash Maps ensure insertion-ordered key-value pairs. URLs resolved by the plugin are
+            // inserted by scheme (key) and a list of that scheme's URL(s) (value)
+            Map<String, List<String>> urlMap = new LinkedHashMap<>();
+
             Optional<ImmutableTriple<String, String, String>> uri = dosPluginUtil.splitURI(targetPath);
 
             if (uri.isPresent() && schemesHandled().contains(uri.get().getLeft())) {
                 Optional<JSONObject> jsonObj = dosPluginUtil.getResponse(uri.get());
 
                 if(jsonObj.isPresent()) {
                     JSONArray urls = jsonObj.get().getJSONObject("data_object").getJSONArray("urls");
+                    // Place all URLs into map for fast lookup time. URLs with duplicate schemes are merged into the existing list
                     for (int i = 0; i < urls.length(); i++) {
-                        urlList.add(urls.getJSONObject(i).getString("url"));
+                        String url = urls.getJSONObject(i).getString("url");
+                        String scheme = url.split(PROTOCOL)[GET_SCHEME];
```

Comment:
There's probably library for this to avoid using a regular expression. 
Perhaps a standard library like https://docs.oracle.com/javase/7/docs/api/java/net/URL.html#getProtocol() or less preferably https://hc.apache.org/httpcomponents-client-ga/httpclient/apidocs/org/apache/http/client/utils/URLEncodedUtils.html#parse(java.lang.String,%20java.nio.charset.Charset)